fv3core/stencils/updatedzc.py

```diff
@@ -1,214 +1,72 @@
 
 
-#      do j=js-ng, je+ng
-#         do i=is-ng, ie+ng
-#            gz2(i,j) = gz(i,j,k)
-#         enddo
-#      enddo
-
-#      if (grid_type < 3) call fill_4corners(gz2, 1, bd, npx, npy, sw_corner, se_corner, ne_corner, nw_corner) # noqa: E501
-
-#      do j=js1, je1
-#         do i=is1, ie2
-#            if( xfx(i,j) > 0.0 ) then
-#                fx(i,j) = gz2(i-1,j)
-#            else
-#                fx(i,j) = gz2(i  ,j)
-#            endif
-#            fx(i,j) = xfx(i,j)*fx(i,j)
-#         enddo
-#      enddo
-
-#      if (grid_type < 3) call fill_4corners(gz2, 2, bd, npx, npy, sw_corner, se_corner, ne_corner, nw_corner) # noqa: E501
-
-#      do j=js1,je2
-#         do i=is1,ie1
-#            if( yfx(i,j) > 0.0 ) then
-#                fy(i,j) = gz2(i,j-1)
-#            else
-#                fy(i,j) = gz2(i,j)
-#            endif
-#            fy(i,j) = yfx(i,j)*fy(i,j)
-#         enddo
-#      enddo
-
-
 @gtscript.function
 def xy_flux(gz_x, gz_y, xfx, yfx):
     fx = xfx * (gz_x[-1, 0, 0] if xfx > 0.0 else gz_x)
     fy = yfx * (gz_y[0, -1, 0] if yfx > 0.0 else gz_y)
     return fx, fy
 
 
-#      do j=js1, je1
-#         do i=is1,ie1
-#            gz(i,j,k) = (gz2(i,j)*area(i,j) +  fx(i,j)- fx(i+1,j)+ fy(i,j)- fy(i,j+1)) & # noqa: E501
-#                      / (         area(i,j) + xfx(i,j)-xfx(i+1,j)+yfx(i,j)-yfx(i,j+1)) # noqa: E501
-#         enddo
-#      enddo
-
-#   end do
-
-# ! Enforce monotonicity of height to prevent blowup
-#   do i=is1, ie1
-#      do j=js1, je1
-#         ws(i,j) = ( zs(i,j) - gz(i,j,km+1) ) * rdt
-#      enddo
-#   end do
-
-#   do k = km, 1, -1
-#      do j=js1, je1
-#         do i=is1, ie1
-#            gz(i,j,k) = max( gz(i,j,k), gz(i,j,k+1) + dz_min )
-#         enddo
-#      enddo
-#   enddo
-
-
 @gtstencil()
-def update_dz_c(
-    dp_ref: sd,
-    zs: sd,
-    area: sd,
-    ut: sd,
-    vt: sd,
-    gz: sd,
-    gz_x: sd,
-    gz_y: sd,
-    ws3: sd,
-    *,
-    dt: float,
+def update_dz_c_stencil(
+    area: FloatField,
+    dp_ref: FloatField,
+    z_surface: FloatField,
+    ut: FloatField,
+    vt: FloatField,
+    gz: FloatField,
+    ws: FloatField,
+    dt2: float,
 ):
+    """Update the model heights from the C-grid wind flux
+
+    After the model runs c_sw and advances the c-grid variables half a timestep,
+    the grid deforms with the flow in the Lagrangian coordinate system. This
+    module updates the model heights based on the flux, and the rate of change of
+    the lowest model height compared to the fixed surface height.
+
+    Args:
+         dp_ref: vertical delta in column reference pressure(in)
+         z_surface: surface height (m) (in)
+         ut: x-velocity on the C-grid, contravariantof the D-grid winds(in)
+         vt: y-velocity on the C-grid, contravariantof the D-grid winds(in)
+         gz: height of the model grid cells (m) (inout)
+         ws: difference in height between lowest model layer and the surface per C-grid
+             timestep(inout)
```

Comment:
Really curious what `ws` is conceptually and why it is calculated. `(gz - gz_surface) / timestep`? Why would that be used? I see on the next line in dyn_core, we throw away everything but the last value in the vertical. It gets passed back to dyn_core where it's called `ws3`, then `riem_solver_c` where it's called `ws` again, then `sim1_solver` where it's called `wsr`, until it's fiiiinally used in one line as `wsr_top`:

```
w = (
    dm * w1 + dt * (pp[0, 0, 1] - pp) - p1 * wsr_top - aa * w[0, 0, -1]
) / bet
```

Crazy thing is, within that function we have everything available to compute ws - we have gz and rdt, and `ws = sum_z(gz/dt)`. Maybe we could move it there in a later refactor? It's such a strange temporary with no real meaning to be moving between all these high-level routines.

Reply:
if possible I'd suggest `surface_delta_gz = gz[lowest level] - gz_surface` with a docstring/comment (I got confused and thought it was `gz[highest level] - gz_surface`, maybe because of the name `wsr_top`). `sim1_solver` already has `rdt` available, so you'd replace `wsr_top` with `surface_delta_gz * rdt`.

`surface_delta_gz_over_dt` is also fine, just a bit of a mouthful, and more complex/confusing than `surface_delta_gz` already is.

As you say we should visit this later and leave the structure for now, but some more thoughts:

> conceptually sort of makes sense it is happening here with the update to gz

I agree it makes a bit of sense, but it's also a little dangerous. `surface_delta_gz` is a diagnostic value, determined by gz (prognostic) and by surface_gz (a constant). If you were to call updatedzc, and then someone added another modification to gz before calling riem_solver, it would introduce a bug because surface_delta_gz would be inaccurate (it would only account for the update from updatedzc and not from the new modification).

Huh, I wonder if the purpose of this is to determine the vertical velocity you need in order to move the lowest level gz to the location of the surface gz in a single timestep, and gets used as the lowest level velocity (in geopotential units) later. That would maybe make sense. If we figure this out we could later name it something like `lowest_level_geopotential_velocity`. For now when we're unsure, something technical and explicit like `surface_delta_gz` is more practical.

Reply:
Given it is called 'ws' I tend to agree, it is likely supposed to be some sort of vertical velocity at the surface, but also agree going with surface_delta_gz for now is good. I think if we move this calculation to the riemann solver, we should do it later, as Eddie has a big lower dimensional storage refactor, so I'd rather make that change after the dust from that settles. Also note there is a parallel with a 'wsd' variables computed in updatedzd and using in riem_solver3 :).